### PR TITLE
Haproxy: add support for draining connections in maintenance mode

### LINKED ISSUE #521: haproxy feature request: wait for sessions to drop to 0 on a specific backend.
##### Issue Type:

“Feature Idea”
##### Ansible Version:

ansible 1.9.1
##### Ansible Configuration:

default
##### Environment:

Debian 7 64bit
##### Summary:

Add haproxy wait for sessions to drop to 0 on a specific backend.
##### Expected Results:

In my case I'm doing rolling upgrades and plan to use the haproxy module to disable specific backends before I upgrade it.

But after I have disabled a backend It would be great to have a command to wait for all sessions to be done, and when 0 connections are using the disabled backend I'll continue with my playbook.

for example.
- wait for backend to reach 0 connections
- timeout, if still connections after X seconds kill connections and exit

```
haproxy: host=backend1 zero_session=yes timeout=30
```

### PR DESCRIPTION
**Pull request by krislindgren**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

modules/network/haproxy
##### ANSIBLE VERSION

```
ansible 2.2.1.0
```

##### SUMMARY

When using drain=yes and wait=yes during disabled the module will
return the action was a success when the server was placed into
maintenance mode and when active sessions have dropped to zero or
until the wait timeout is reached, as long as the server is in
maintenance mode.

Fixes: https://github.com/ansible/ansible-modules-extras/issues/521

```

```

### CODE REVIEW

**Comment by harlowja** [comment]
What's 'harpoxy' ;)

File: lib/ansible/modules/network/haproxy.py [hunk 1]
```diff
@@ -298,16 +321,30 @@ def wait_until_status(self, pxname, svname, status):
...
+        in_correct_state = False
         for i in range(1, self.wait_retries):
             state = self.get_state_for(pxname, svname)
 
             # We can assume there will only be 1 element in state because both svname and pxname are always set when we get here
             if state[0]['status'] == status:
-                return True
+                in_correct_state = True
+                if self.drain and status == 'MAINT':
+                    #we are in main, now check if the we are currently at 0 sessions
```
Comment (harlowja) [inline lib/ansible/modules/network/haproxy.py]: Follow pep8 for block comments.

`Each line of a block comment starts with a # and a single space`

**Comment by harlowja** [comment]
Does this implement https://github.com/ansible/ansible-modules-extras/issues/521 ?

File: lib/ansible/modules/network/haproxy.py [hunk 2]
```diff
@@ -287,8 +307,11 @@ def get_state_for(self, pxname, svname):
         """
         data = self.execute('show stat', 200, False).lstrip('# ')
         r = csv.DictReader(data.splitlines())
-        state = tuple(map(lambda d: { 'status': d['status'], 'weight': d['weight'] }, filter(lambda d: (pxname is None or d['pxname'] == pxname) and d['svname'] == svname, r)))
-        return state or None
+        state = []
+        for d in r:
+            if (pxname is None or d['pxname'] == pxname) and d['svname'] == svname:
+                state.append({ 'status': d['status'], 'weight': d['weight'], 'scur': d['scur'] })
+        return state
```
Comment (harlowja) [inline lib/ansible/modules/network/haproxy.py]: Not sure if it matters but prior code had `return state or None` which will now not be `None`; might not matter, but it's a functional difference.

Comment (krislindgren) [reply lib/ansible/modules/network/haproxy.py]: Done.

File: lib/ansible/modules/network/haproxy.py [hunk 3]
```diff
@@ -42,6 +42,14 @@
       - Name of the HAProxy backend pool.
     required: false
     default: auto-detected
+  drain:
+    description:
+      - Wait until the server has no active connections or until timeout from
+        wait_interval, wait_retries is reached.  At which point as long as we
+        are in status of 'MAINT', we will continue.
+    required: false
```
Comment (bjolivot) [inline lib/ansible/modules/network/haproxy.py]: No need to have the `required: false` lines, it's implicit in documentation

File: lib/ansible/modules/network/haproxy.py [hunk 4]
```diff
@@ -378,6 +415,7 @@ def main():
             wait=dict(required=False, default=False, type='bool'),
             wait_retries=dict(required=False, default=WAIT_RETRIES, type='int'),
             wait_interval=dict(required=False, default=WAIT_INTERVAL, type='int'),
+            drain=dict(required=False, default=False, type='bool'),
```
Comment (bjolivot) [inline lib/ansible/modules/network/haproxy.py]: No need to have the required= False it's implicit in documentation module spec.

**Review by bjolivot — COMMENTED** [review]
Some unneeded lines

File: lib/ansible/modules/network/haproxy.py (hunk 4 above)
Comment (krislindgren) [reply lib/ansible/modules/network/haproxy.py]: Just following whats the rest of the module was doing.

File: lib/ansible/modules/network/haproxy.py (hunk 3 above)
Comment (krislindgren) [reply lib/ansible/modules/network/haproxy.py]: Just following other examples in the module.

**Comment by gundalow** [comment]
Hi,
A number of files have been git moved from `modules/network` to `modules/net_tools`, which has effected this PR. Please rebase this PR to pull in that change.

**Comment by gundalow** [comment]
rebase needed

**Comment by drmrd** [comment]
@krislindgren Are you still working on this PR? I'm interested in seeing this feature added, and it seems like your PR has gone stale since ansible-modules-extra was merged into this repo. I'll be creating a newly rebased PR (retaining your original commit) and linking to it from here (and vice versa). If you release an updated version of your PR, I'll be happy to remove mine. Thanks for implementing this feature!

**Comment by drmrd** [comment]
I've reproduced this merge request here: #25887

**Comment by gundalow** [comment]
Given we've not heard anything since 19th May I wonder if this should be closed and we should continue in  https://github.com/ansible/ansible/pull/25887

**Pull request closed without merging**